File: server/src/validation.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { PERSONALITY_DIR, readAllPersonalityFiles } from "./personality.js";
import { resolveSkillsDir } from "./skills.js";
import type { CrossRefCheck, CrossRepoCheck } from "./types.js";

function extractHeadings(body: string): string[] {
  const anchors: string[] = [];
  for (const line of body.split("\n")) {
    const m = line.match(/^## (.+)/);
    if (m) {
      anchors.push(m[1].toLowerCase().replace(/\s+/g, "-"));
    }
  }
  return anchors;
}

export function validateCrossReferences(): CrossRefCheck[] {
  const files = readAllPersonalityFiles();
  const anchorMap = new Map<string, string[]>();
  for (const f of files) {
    anchorMap.set(f.filename, extractHeadings(f.body));
  }
  const results: CrossRefCheck[] = [];
  for (const f of files) {
    for (const ref of f.frontmatter.crossReferences) {
      const m = ref.match(/^(\w+\.md)#(\S+)\s*--\s*(.+)$/);
      if (!m) {
        results.push({
          sourceFile: f.filename,
          targetFile: "?",
          targetAnchor: "?",
          description: ref,
          valid: false,
          error: "Malformed cross-reference",
        });
        continue;
      }
      const [, targetFile, targetAnchor, desc] = m;
      const headings = anchorMap.get(targetFile);
      if (!headings) {
        results.push({
          sourceFile: f.filename,
          targetFile,
          targetAnchor,
          description: desc,
          valid: false,
          error: `File ${targetFile} not found`,
        });
      } else if (!headings.includes(targetAnchor)) {
        results.push({
          sourceFile: f.filename,
          targetFile,
          targetAnchor,
          description: desc,
          valid: false,
          error: `Anchor #${targetAnchor} not found in ${targetFile}`,
        });
      } else {
        results.push({
          sourceFile: f.filename,
          targetFile,
          targetAnchor,
          description: desc,
          valid: true,
        });
      }
    }
  }
  return results;
}

export function validateCrossRepoReferences(options?: {
  personalityDir?: string;
  skillsDir?: string;
}): CrossRepoCheck[] {
  const pDir = options?.personalityDir ?? PERSONALITY_DIR;
  const sDir = options?.skillsDir ?? resolveSkillsDir();
  const results: CrossRepoCheck[] = [];

  const pFiles = fs.readdirSync(pDir).filter((f) => f.endsWith(".md"));
  for (const pf of pFiles) {
    const raw = fs.readFileSync(path.join(pDir, pf), "utf-8");
    const skillRefs = raw.matchAll(/skills:(\w+)/g);
    for (const m of skillRefs) {
      const skillName = m[1];
      const skillPath = path.join(sDir, skillName, "SKILL.md");
      const exists = fs.existsSync(skillPath);
      results.push({
        source: `personality:${pf}`,
        target: `skills:${skillName}`,
        valid: exists,
        error: exists ? undefined : `Skill "${skillName}" not found in skills repo`,
      });
    }
  }

  if (fs.existsSync(sDir)) {
    const skillDirs = fs.readdirSync(sDir, { withFileTypes: true });
    for (const d of skillDirs) {
      if (!d.isDirectory() || d.name.startsWith(".")) continue;
      const skillMd = path.join(sDir, d.name, "SKILL.md");
      if (!fs.existsSync(skillMd)) continue;
      const raw = fs.readFileSync(skillMd, "utf-8");
      const pRefs = raw.matchAll(/personality:(\w+\.md)/g);
      for (const m of pRefs) {
        const pFile = m[1];
        const pPath = path.join(pDir, pFile);
        const exists = fs.existsSync(pPath);
        results.push({
          source: `skills:${d.name}/SKILL.md`,
          target: `personality:${pFile}`,
          valid: exists,
          error: exists ? undefined : `Personality file "${pFile}" not found`,
        });
      }
    }
  }

  return results;
}

File: server-py/src/ai_personality/validation.py
import re
from pathlib import Path

from .personality import read_all_personality_files, get_personality_dir
from .skills import resolve_skills_dir


def _extract_headings(body: str) -> list[str]:
    return [
        m.group(1).lower().replace(" ", "-")
        for m in re.finditer(r"^## (.+)", body, re.MULTILINE)
    ]


def validate_cross_references() -> list[dict]:
    files = read_all_personality_files()
    anchor_map: dict[str, list[str]] = {}
    for f in files:
        anchor_map[f["filename"]] = _extract_headings(f["body"])

    results = []
    for f in files:
        for ref in f["frontmatter"].get("crossReferences", []):
            m = re.match(r"^(\w+\.md)#(\S+)\s*--\s*(.+)$", ref)
            if not m:
                results.append(
                    {
                        "sourceFile": f["filename"],
                        "targetFile": "?",
                        "targetAnchor": "?",
                        "description": ref,
                        "valid": False,
                        "error": "Malformed cross-reference",
                    }
                )
                continue
            target_file, target_anchor, desc = m.groups()
            headings = anchor_map.get(target_file)
            if headings is None:
                results.append(
                    {
                        "sourceFile": f["filename"],
                        "targetFile": target_file,
                        "targetAnchor": target_anchor,
                        "description": desc,
                        "valid": False,
                        "error": f"File {target_file} not found",
                    }
                )
            elif target_anchor not in headings:
                results.append(
                    {
                        "sourceFile": f["filename"],
                        "targetFile": target_file,
                        "targetAnchor": target_anchor,
                        "description": desc,
                        "valid": False,
                        "error": f"Anchor #{target_anchor} not found in {target_file}",
                    }
                )
            else:
                results.append(
                    {
                        "sourceFile": f["filename"],
                        "targetFile": target_file,
                        "targetAnchor": target_anchor,
                        "description": desc,
                        "valid": True,
                    }
                )
    return results


def validate_cross_repo_references(
    personality_dir: Path | str | None = None,
    skills_dir: Path | str | None = None,
) -> list[dict]:
    p_dir = Path(personality_dir) if personality_dir else get_personality_dir()
    s_dir = Path(skills_dir) if skills_dir else resolve_skills_dir()
    results = []

    if p_dir.exists():
        for pf in p_dir.iterdir():
            if not pf.name.endswith(".md") or not pf.is_file():
                continue
            raw = pf.read_text("utf-8")
            for m in re.finditer(r"skills:(\w+)", raw):
                skill_name = m.group(1)
                skill_path = s_dir / skill_name / "SKILL.md"
                exists = skill_path.exists()
                results.append(
                    {
                        "source": f"personality:{pf.name}",
                        "target": f"skills:{skill_name}",
                        "valid": exists,
                        "error": (
                            None
                            if exists
                            else f'Skill "{skill_name}" not found in skills repo'
                        ),
                    }
                )

    if s_dir.exists():
        for entry in s_dir.iterdir():
            if not entry.is_dir() or entry.name.startswith("."):
                continue
            skill_md = entry / "SKILL.md"
            if not skill_md.exists():
                continue
            raw = skill_md.read_text("utf-8")
            for m in re.finditer(r"personality:(\w+\.md)", raw):
                p_file = m.group(1)
                p_path = p_dir / p_file
                exists = p_path.exists()
                results.append(
                    {
                        "source": f"skills:{entry.name}/SKILL.md",
                        "target": f"personality:{p_file}",
                        "valid": exists,
                        "error": (
                            None if exists else f'Personality file "{p_file}" not found'
                        ),
                    }
                )

    return results
